File: patterns/orchestration/state-checkpointing/src/ts/stores.ts
import { CheckpointStore, WorkflowCheckpoint } from "./types.js";

/**
 * In-memory checkpoint store for testing and development.
 *
 * Does NOT survive process restarts. For production, replace with a
 * RedisCheckpointStore or PostgresCheckpointStore that persists to durable storage.
 * The interface is the same — the store is the only thing that changes.
 */
export class InMemoryCheckpointStore implements CheckpointStore {
  private store = new Map<string, WorkflowCheckpoint>();

  async save(workflowId: string, checkpoint: WorkflowCheckpoint): Promise<void> {
    // Deep clone on write to prevent external mutation of stored state.
    // In a real store this is handled by serialization (JSON to Redis, etc.)
    this.store.set(workflowId, JSON.parse(JSON.stringify(checkpoint)));
  }

  async load(workflowId: string): Promise<WorkflowCheckpoint | null> {
    const checkpoint = this.store.get(workflowId);
    if (!checkpoint) return null;
    // Return a clone to prevent callers from mutating internal state
    return JSON.parse(JSON.stringify(checkpoint));
  }

  async clear(workflowId: string): Promise<void> {
    this.store.delete(workflowId);
  }

  // Utility for tests: inspect store size without exposing internals
  size(): number {
    return this.store.size;
  }

  // Utility for tests: list all workflow IDs in the store
  keys(): string[] {
    return Array.from(this.store.keys());
  }
}

File: patterns/orchestration/state-checkpointing/src/py/stores.py
from __future__ import annotations

import copy

from .types import CheckpointStore, WorkflowCheckpoint


class InMemoryCheckpointStore(CheckpointStore):
    """
    In-memory checkpoint store for testing and development.

    Does NOT survive process restarts. For production, replace with a
    RedisCheckpointStore or PostgresCheckpointStore. The interface is
    identical — only the store implementation changes.
    """

    def __init__(self) -> None:
        self._store: dict[str, WorkflowCheckpoint] = {}

    async def save(self, workflow_id: str, checkpoint: WorkflowCheckpoint) -> None:
        # Deep copy on write to prevent external mutation of stored state.
        # In a real store this is handled by serialization (to Redis, Postgres, etc.)
        self._store[workflow_id] = copy.deepcopy(checkpoint)

    async def load(self, workflow_id: str) -> WorkflowCheckpoint | None:
        checkpoint = self._store.get(workflow_id)
        if checkpoint is None:
            return None
        return copy.deepcopy(checkpoint)

    async def clear(self, workflow_id: str) -> None:
        self._store.pop(workflow_id, None)

    def size(self) -> int:
        return len(self._store)

    def keys(self) -> list[str]:
        return list(self._store.keys())
